swarm/storage/mru/lookup/lookup.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+/*
+Package lookup defines resource lookup algorithms and provides tools to place updates
+so they can be found
+*/
+package lookup
+
+const maxuint64 = ^uint64(0)
+const lowestLevel uint8 = 0
+const numLevels uint8 = 26
+const highestLevel = lowestLevel + numLevels - 1
+const defaultLevel = highestLevel
+
+// ReadFunc is a handler called by Lookup each time it attempts to find a value
+// It should return <nil> if a value is not found
+// It should return <nil> if a value is found, but its timestamp is higher than "now"
+// It should only return an error in case the handler wants to stop the
+// lookup process entirely.
+type ReadFunc func(epoch Epoch, now uint64) (interface{}, error)
+
+// NoClue is a hint that can be provided when the Lookup caller does not have
+// a clue about where the last update may be
+var NoClue = Epoch{}
+
+func getBaseTime(t uint64, level uint8) uint64 {
+	return t & (maxuint64 << level)
+}
+
+// Hint creates a hint based only on the last known update time
+func Hint(last uint64) Epoch {
+	return Epoch{
+		Time:  last,
+		Level: defaultLevel,
+	}
+}
+
+func getNextLevel(last Epoch, now uint64) uint8 {
```

Comment:
Please add a brief description of the algorithm here (maybe just cut and paste from the PR description)

Furthermore, the algorithm doesn't seem to have tests. I would suggest having exhaustive tests to confirm it yields sane values.

Reply:
Added a brief description and inline comments

Regarding tests, the `getNextLevel()` algorithm is actually simple, although the bit shuffling obfuscates it a bit. Added comments to clarify. 
It is also fully covered by the package tests. This is a private function and not accessible from the test package.

The whole package has a coverage of 93%

Reply:
> the bit shuffling obfuscates it a bit

Yes I know it's private, and I know the code is being used.

But as far as I can see you're not testing anywhere whether the function returns the value you would expect of it. It would not hurt to test that.

Reply:
ok, done. Wrote a comprehensive test to beat the *@#! out of `GetNextLevel()`.

Reply:
Thank you for obliging me! :field_hockey:

Reply:
Where did you put it?